what4/src/What4/Utils/BVDomain.hs

```diff
@@ -1,18 +1,503 @@
 {-|
 Module      : What4.Utils.BVDomain
-Copyright   : (c) Galois Inc, 2015-2016
+Copyright   : (c) Galois Inc, 2019
 License     : BSD3
-Maintainer  : jhendrix@galois.com
+Maintainer  : huffman@galois.com
 
-Provides an implementation of bitvector abstract domains.
-This module reexports either the "empty" implementation,
-which does essentially no work, or the "maps" implementation.
-This setup allows easy compile-time selection of an implementation,
-for performance comparison purposes.
+Provides an interval-based implementation of bitvector abstract
+domains.
 -}
 
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeOperators #-}
 
-module What4.Utils.BVDomain (module BVDomain) where
+module What4.Utils.BVDomain
+  ( BVDomain
+    -- * Projection functions
+  , asSingleton
+  , ubounds
+  , sbounds
+  , eq
+  , slt
+  , ult
+  , testBit
+  , domainsOverlap
+  , ranges
+    -- * Operations
+  , any
+  , singleton
+  , range
+  , fromAscEltList
+  , union
+  , concat
+  , select
+  , zext
+  , sext
+    -- ** Shifts
+  , shl
+  , lshr
+  , ashr
+    -- ** Arithmetic
+  , add
+  , negate
+  , scale
+  , mul
+  , udiv
+  , urem
+  , sdiv
+  , srem
+    -- ** Bitwise
+  , What4.Utils.BVDomain.not
+  , and
+  , or
+  , xor
+  ) where
 
+import qualified Data.Bits as Bits
+import           Data.Bits hiding (testBit, xor)
+import           Data.Parameterized.NatRepr
+import           Numeric.Natural
+import           GHC.TypeNats
+import           GHC.Stack
 
-import What4.Utils.BVDomain.Map as BVDomain
+import           Prelude hiding (any, concat, negate, and, or)
+
+--------------------------------------------------------------------------------
+-- BVDomain definition
+
+-- | A value of type @'BVDomain' w@ represents a set of bitvectors of
+-- width @w@. Each 'BVDomain' can represent a single contiguous
+-- interval of bitvectors that may wrap around from -1 to 0.
+data BVDomain (w :: Nat)
+  = BVDAny !Integer
+  -- ^ The set of all bitvectors of width @w@. Argument caches @2^w-1@.
+  | BVDInterval !Integer !Integer !Integer
+  -- ^ Intervals are represented by a starting value and a size.
+  -- @BVDInterval l d@ represents the set of values of the form @x mod
```

Comment:
Is the first argument @2^w-1@?  If so, maybe write `@BVInterval (2^w-1) l d@` instead of @BVDInterval l d@?

Reply:
Also would any performance be gained by using `Natural` instead of `Integer`?

Reply:
Yes, I wrote that comment before I added the cached bitmask; I'll fix it. Also, I thought about using `Natural` to enforce non-negativity, but I don't think you'd get a performance benefit, as the runtime representation for `Natural` is almost identical to that for `Integer`. But then you'd have to add some casts between `Natural` and `Integer` in various places which would clutter the code. And we're already relying on data abstraction to maintain stronger datatype invariants, so switching to `Natural` wouldn't buy us much there.

Reply:
Thanks for the additional information.  One question about this new representation.  In machine code, I expect that we'll see a lot of cases where code is doing say 32-bit signed arithmetic and then storing the value in a 64-bit register.  I may misremember the old code, but if we had a range constraint with a negative lower bound `[-1.. 3]` on the 32-bit value, I think the old representation would use two ranges, while this would use one, and the 64-bit value constraint would be very imprecise.
I think in practice this won't really matter as the 64-bit stored value would be truncated, and we should have rewrite rules to simplify away the extension/truncation, but it may be good to do a bit of testing or thinking about this problem, because it is a very common occurrence in machine code.

Reply:
Comment fixed in revision bac273fc031181be543e5a7c6e41085a30c39c1a.